Sources/SourceKitLSP/SourceKitLSPServer.swift

```diff
@@ -1403,9 +1403,26 @@ extension SourceKitLSPServer {
   }
 
   func didChangeActiveDocument(_ notification: DidChangeActiveDocumentNotification) async {
-    let workspace = await self.workspaceForDocument(uri: notification.textDocument.uri)
-    await workspace?.semanticIndexManager?
-      .schedulePreparationForEditorFunctionality(of: notification.textDocument.uri)
+    guard let activeDocument = notification.textDocument?.uri else {
+      // The client no longer has a SourceKit-LSP document open. Mark in-progress preparations as irrelevant to cancel
+      // them if they haven't started yet.
+      for workspace in workspaces {
+        await workspace.semanticIndexManager?.markPreparationForEditorFunctionalityTaskAsIrrelevant()
+      }
+      return
+    }
+    guard self.documentManager.openDocuments.contains(activeDocument) else {
+      logger.error("Trying to mark \(activeDocument) as active document but document is not open")
```

Comment:
Does this mean `DidChangeActiveDocumentNotification` can no longer be `.freestanding` since it can race with `DidOpenTextDocumentNotification`?